pkg/controller/volume/attachdetach/attach_detach_controller.go

```diff
@@ -475,6 +475,58 @@ func (adc *attachDetachController) populateDesiredStateOfWorld() error {
 		}
 	}
 
+	// Process Volume-Attachment objects.
+	// If the volume in a VA object is not present in the ASW, add the volume
+	// to ASW as an "uncertain" attachment. If the volume is present in the DSW,
+	// the reconciler will attempt attach on the volume, if the volume is NOT
+	// present in the DSW (could be a dangling attachment), the reconciler will
+	// detach this volume.
+	vas, err := adc.volumeAttachmentLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("failed to list VolumeAttachment objects: %v", err)
+		return err
+	}
+	for _, va := range vas {
+		nodeName := types.NodeName(va.Spec.NodeName)
+		pvName := va.Spec.Source.PersistentVolumeName
+		if pvName == nil {
+			klog.V(5).Infof("Skipping the va as its pvName is nil, va.Name: %q, nodeName: %q",
+				va.Name, nodeName)
+			continue
+		}
+		pv, err := adc.pvLister.Get(*pvName)
+		if err != nil {
+			klog.Errorf("Unable to lookup pv object for: %q, err: %v", *pvName, err)
+			continue
+		}
+		volumeSpec := volume.NewSpecFromPersistentVolume(pv, false)
+		plugin, err := adc.volumePluginMgr.FindAttachablePluginBySpec(volumeSpec)
+		if err != nil || plugin == nil {
+			klog.V(5).Infof(
+				"Skipping processing the volume %q on nodeName: %q, no attacher interface found. err=%v",
+				*pvName,
+				nodeName,
+				err)
+			continue
+		}
+		volumeName, err := volumeutil.GetUniqueVolumeNameFromSpec(plugin, volumeSpec)
+		if err != nil {
+			klog.Errorf(
+				"Failed to find unique name for volume:%q, va.Name:%q, nodeName:%q: %v",
+				*pvName,
+				va.Name,
+				nodeName,
+				err)
+			continue
+		}
+		if !adc.actualStateOfWorld.IsVolumeAttachedToNode(volumeName, nodeName) {
```

Comment:
I am thinking to remove this, putting into uncertain state is even safer considering the attach status might not be the true status.

Reply:
also I am thinking it is better to mark as uncertain when processing from node status volumeAttached list.

Reply:
Sorry, I didnt understand. Do you mean add volume to uncertain unconditionally? That means we will move all volumes to uncertain. I dont think we want that.
Note that during attach, volume is added to node->VolumesAttached when the transaction ends, and during detach, the volume is removed from node->VolumesAttached when the transaction begins. So, if we have a VA object, but the volume is not in VolumesAttached that means it is a dangling-attach candidate, so we mark as uncertain.

Reply:
Since this is part of populate actual state when controller restarts, it is triggered before reconciler starts to work. At this point, the controller first to use node status to update asw, and you are adding here to use VA to further update asw. Because controller is restarting, it might be possible that node status volumeattached list is not representing the real state. For example, node is deleted and recreated while node object still exist. So volume is automatically detached, but it is. still on volumeAttached list..

I think putting uncertain will make sure that attach/detach will be triggered which is safer. Controller might trigger unnecessary attach/detach, but I think it is ok. It is better cover all corner cases with a little bit overhead.

Reply:
>> For example, node is deleted and recreated while node object still exist. 
When a node is deleted, isn't the node object deleted? Are you saying the node deletion happened after asw processed the node, but before the node->event handler is setup? If so, then this volume wont be in DSW (as no pods are running on the node), so will be detached and cleared? Isnt it? Are you saying a dangling attach is possible?

Anyways, I think marking every VA as uncertain will lead to issuing attach for every good volume! I still think this may be undesired as there could be other implications of doing this.

Reply:
Discussed offline with jingxu97@ and msau@. The conclusion is to create a separate issue to investigate the need for marking all the volumes as 'uncertain' and potential implications of such a change. We discussed that it may result in cloud-provider attach or get calls for all good (intree) volumes too. Theoretically, attach() should be idempotent, but could be risky in some drivers. We need to validate these risks and understand the scale implications. Also, if a node object remains while the underlying machine is recreated, the ADC's periodic VerifyVolumesAreAttached() would potentially detect the volumes are not attached to this machine and fix it.  jingxu97@ will create an issue to followup on this. Thanks!